### ISSUE TITLE
`partial_double` and `_init` with parameters

### ISSUE THREAD

**Issue by RoniPerson** [bitwes/Gut#354]
I have a class that takes some setup parameters in `_init`. When I tried doubling it with `partial_double` I got an error since `new` would not accept the setup parameters.
```var inst = partial_double('res://path/to/my/class.gd').new(true)```
Did I miss something in the documentation or is this not possible right now?

**Comment by bitwes** [comment]
It will work if you add default values for all of the _init parameters.

**Comment by RoniPerson** [comment]
Yes it totally works as long as I don't pass any parameters. But what do I do when I want to test the object with special initialisation (diffrent from the default values)?

**Comment by bitwes** [comment]
This is not supported (yet).  The workaround would be to create initialization methods that `_init` and your test can call.  `_init` is always called when you create an instance, there's no way to get around that.  So you have to be sure that your initialization methods can be called twice.  There will be one call when you call `.new()` which will get defaults, and then the call you make in your test.

## The fix
GUT uses metadata about a class, provided by Godot, to find all methods, signals, parameters, etc for that class.  Previously I could not find any case where the metadata listed parameters for the `_init` method.  I'm not sure if this changed or I never noticed that there are two listings for `_init` in the metadata.  I'm 90% sure this wasn't always the case.

I created 
```
class InitParameters:
	var a = 'b'

	func _init(value='asdf'):
		a = value
```
and in the results of `get_method_list` I found two entries for `_init`:
```
{args:[], default_args:[], flags:33, id:0, name:_init, return:{class_name:, hint:0, hint_string:, name:, type:0, usage:7}}, 
```
and 
```
{args:[{class_name:, hint:0, hint_string:, name:arg0, type:0, usage:7}], default_args:[], flags:65, id:0, name:_init, return:{class_name:, hint:0, hint_string:, name:, type:0, usage:7}}, 
```
The 2nd entry has `flags:65` which indicates a local overload of the method.  It also has the parameter listed.  This means this can be fixed now.

**Comment by bitwes** [comment]
There's also the issue where you have to explicitly make your _init methods send parameters to super.  I originally only thought this was important if you wanted to change the number of parameters, but it appears to be necessary whenever `_init` has parameters and you want to pass the same values upstream.

Given:
```gdscript
class Thing1:
    var value = 'thing1_start'
    func _init(new_val='thing1_default'):
        print('thing 1 got ', new_val)
        value = new_val

class Thing2:
    extends Thing1

    func _init(new_val='thing2_default'):
        print("thing 2 got ", new_val)

func before_all():
    gut.get_doubler()._print_source = true

func after_all():
    gut.get_doubler()._print_source = false

func test_nothing():
    var t1_1 = Thing1.new()
    print('t1_1: ', t1_1.value, "\n")

    var t1_2 = Thing1.new('made it')
    print('t1_2: ', t1_2.value, "\n")

    var t2_1 = Thing2.new()
    print('t2_1: ', t2_1.value, "\n")

    var t2_2 = Thing2.new('made again')
    print('t2_2: ', t2_2.value, "\n")
```
outputs 
```
thing 1 got thing1_default
t1_1: thing1_default

thing 1 got made it
t1_2: made it

thing 1 got thing1_default
thing 2 got thing2_default
t2_1: thing1_default

thing 1 got thing1_default
thing 2 got made again
t2_2: thing1_default
```

But if you change `Thing2`'s `_init` to:
```gdscript
func _init(new_val='thing 2 default').(new_val):
```
You get 
```
thing 1 got thing1_default
t1_1: thing1_default

thing 1 got made it
t1_2: made it

thing 1 got thing2_default
thing 2 got thing2_default
t2_1: thing2_default

thing 1 got made again
thing 2 got made again
t2_2: made again
```

**Comment by bitwes** [comment]
This has been implemented in master in PR #359 and will be in the next release.

You can now pass parameters to `new` when making instances of doubles and partial doubles.  You can also stub default parameter values (as they will get `null` as a default) and spy on values passed to `_init` through `new(...)`.
